src/trio/_core/_tests/test_run.py

```diff
@@ -3005,3 +3005,27 @@ async def main() -> None:
             await _core.wait_task_rescheduled(inner_abort)
 
     _core.run(main)
+
+
+async def test_closed_task_iter_await_frames() -> None:
+    """Test that Task.__iter__ handles cr_frame/gi_frame becoming None.
+
+    When a coroutine completes, its cr_frame/gi_frame attributes become None.
+    The iterator should handle this gracefully rather than trying to access
+    attributes on None.
+    """
+    completed_task = None
+
+    async with _core.open_nursery() as nursery:
+
+        async def capture_task() -> None:
+            nonlocal completed_task
+            completed_task = _core.current_task()
+            await _core.checkpoint()
+
+        nursery.start_soon(capture_task)
+
+    # Task has completed, so coro.cr_frame should be None, thus no frames
+    assert completed_task is not None
```

Comment:
```suggestion
    assert completed_task.coro is not None
```

Also I thought for a bit `coro` would be `None` once a `Task` finished and that that was the TOC/TOU bug you were describing :-)